tensorflow/compiler/xla/service/gpu/cudnn_conv_runner.cc

```diff
@@ -73,31 +73,80 @@ class ScratchBufAllocator : public se::ScratchAllocator {
   bool allocated_ = false;
 };
 
-template <typename T>
```

Comment:
There are many code changes in this file, mostly for one refactoring.

The refactoring starts from RunCudnnConvImpl.  RunCudnnConvImpl used to dispatch for convolution kind and wasn't necessary to dispatch for data types.  We now have to dispatch for data types for integer convolutions, because CuDNN integer convolution have multiple input/output type combinations and doesn't support all kinds of convolutions.

After refactoring, RunCudnnConvImpl calls RunCudnnConvInternalImpl specializations to dispatch for data types.  Each RunCudnnConvInternalImpl specialization dispatches for convolution kind and only calls supported ones.

Reply:
Sure, I will.